sklearn/linear_model/_glm/tests/test_glm.py

```diff
@@ -29,6 +64,480 @@ def regression_data():
+    assert model.intercept_ == pytest.approx(intercept, rel=rtol)
+    assert_allclose(model.coef_, coef, rtol=rtol)
+
+
+@pytest.mark.parametrize("solver", SOLVERS)
+@pytest.mark.parametrize("fit_intercept", [True, False])
+def test_glm_regression_unpenalized(solver, fit_intercept, glm_dataset):
+    """Test that unpenalized GLM converges for all solvers to correct solution.
+
+    We work with a simple constructed data set with known solution.
+    Note: This checks the minimum norm solution for wide X, i.e.
+    n_samples < n_features:
+        min ||w||_2 subject to w = argmin deviance(X, y, w)
+    """
+    model, X, y, coef, _, _ = glm_dataset
+    n_samples, n_features = X.shape
+    alpha = 0  # unpenalized
+    params = dict(
+        alpha=alpha,
+        fit_intercept=fit_intercept,
+        # solver=solver,  # only lbfgs available
+        tol=1e-12,
+        max_iter=1000,
+    )
+
+    model = clone(model).set_params(**params)
+    if fit_intercept:
+        X = X[:, :-1]  # remove intercept
+        intercept = coef[-1]
+        coef = coef[:-1]
+    else:
+        intercept = 0
+
+    if n_samples < n_features:
+        warnings.filterwarnings("ignore", category=scipy.linalg.LinAlgWarning)
+        warnings.filterwarnings("ignore", category=ConvergenceWarning)
+    model.fit(X, y)
+
+    # FIXME: `assert_allclose(model.coef_, coef)` should work for all cases but fails
+    # for the wide/fat case with n_features > n_samples. Most current GLM solvers do
+    # NOT return the minimum norm solution with fit_intercept=True.
+    rtol = 5e-5
+    if n_samples > n_features:
+        assert model.intercept_ == pytest.approx(intercept)
+        assert_allclose(model.coef_, coef, rtol=rtol)
+    else:
+        # As it is an underdetermined problem, prediction = y. The following shows that
+        # we get a solution, i.e. a (non-unique) minimum of the objective function ...
+        assert_allclose(model.predict(X), y, rtol=1e-6)
+        if fit_intercept:
+            # But it is not the minimum norm solution. Otherwise the norms would be
+            # equal.
+            norm_solution = np.linalg.norm(np.r_[intercept, coef])
+            norm_model = np.linalg.norm(np.r_[model.intercept_, model.coef_])
+            assert norm_model > (1 + 1e-12) * norm_solution
+
+            # Note: Even adding a tiny penalty does not give the minimal norm solution.
+            # model_pen = clone(model).set_params(**params).set_params(alpha=1e-10)
+            # model_pen.fit(X, y)
+            # assert_allclose(model_pen.predict(X), y, rtol=1e-6)  # This is true.
+            # norm_model_pen = np.linalg.norm(
+            #     np.r_[model_pen.intercept_, model_pen.coef_]
+            # )
+            # All the following assertions fail.
+            # assert norm_model_pen == pytest.approx(norm_solution)
+            # assert model_pen.intercept_ == pytest.approx(intercept)
+            # assert_allclose(model_pen.coef_, coef, rtol=rtol)
```

Comment:
what do we do with this? keep this for later hence adding XXX?
or we cleanup and just remove this?

Reply:
I'll add an XXX as I think having these pieces of code here helps to know exactly what fails.

Reply:
We could make it a `XXX` as I originally thought in https://github.com/scikit-learn/scikit-learn/pull/23619#pullrequestreview-1009473106 but then I thought I wasn't sure it was a bug or not. I don't know if we can expect that LBFGS should converge to the minimal norm solution in the limit of tiny penalty or how to configure the relative tolerances of the various internal stopping criteria of LBFGS to guarantee this.

That being said, I also don't link commented out code in the code base. Maybe we should instead extract this as a reproducer snippet and put it in an issue instead (and just reference the issue in a comment in the test).

Reply:
I've opend an issue #23670.